### PR TITLE
Fix incorrect readbehind and try to make readahead clearer and easier.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build  Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Review?

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -6680,6 +6681,62 @@ ensure_method_is_allowed_to_call_method (MonoCompile *cfg, MonoMethod *caller, M
...
+			break;
+		}
+		// FIXME verify it is within the function.
+		*target = ip + delta;
+		return ip;
+	}
+	return NULL;
+}
+
+#define il_read_brtrue(ip, end, target) 	(il_read_branch_and_target (ip, end, CEE_BRTRUE,    MONO_CEE_BRTRUE,    4, target))
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Ultimately these should be more data driven, but this is a start.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

File: mono/mini/method-to-ir.c [hunk 2]
```diff
@@ -6680,6 +6681,62 @@ ensure_method_is_allowed_to_call_method (MonoCompile *cfg, MonoMethod *caller, M
...
+static unsigned char*
+il_read_op (unsigned char *ip, unsigned char *end, unsigned char first_byte, MonoOpcodeEnum desired_il_op)
+// If ip is desired_il_op, return the next ip, else NULL.
+{
+	if (G_LIKELY (ip < end) && G_UNLIKELY (*ip == first_byte)) {
+		MonoOpcodeEnum il_op = MonoOpcodeEnum_Invalid;
+		// mono_opcode_value_and_size updates ip, but not in the expected way.
+		const unsigned char *temp_ip  = ip;
+		const int size = mono_opcode_value_and_size (&temp_ip, end, &il_op);
+		return G_UNLIKELY (size > 0 && il_op == desired_il_op) ? (ip + size) : NULL;
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: This  should be likely (size > 0) and unlikely(ip == desired).

File: mono/metadata/opcodes.h [hunk 3]
```diff
@@ -23,10 +23,15 @@ MONO_BEGIN_DECLS
 #define OPDEF(a,b,c,d,e,f,g,h,i,j) \
 	MONO_ ## a,
 
-typedef enum {
+#ifndef __MONO_METADATA_BASIC_BLOCK_H__
```
Comment (kumpera) [inline mono/metadata/opcodes.h]: What's so broken with mono that we need this?

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -8126,36 +8181,41 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
 			}
 			*sp++ = ins;
 			break;
+
 		case CEE_LDLOC_0:
 		case CEE_LDLOC_1:
 		case CEE_LDLOC_2:
 		case CEE_LDLOC_3:
+			n = (*ip) - CEE_LDLOC_0;
+ldloc:
```
Comment (kumpera) [inline mono/mini/method-to-ir.c]: I'm not a fan of local labels inside switch cases. Maybe we can later move this out to a function?

File: mono/mini/method-to-ir.c [hunk 5]
```diff
@@ -11689,9 +11722,9 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 
 				EMIT_NEW_TEMPLOAD (cfg, ins, vtvar->inst_c0);
 			} else {
-				if ((ip + 9 < end) && ip_in_bb (cfg, cfg->cbb, next_ip) &&
+				if ((next_ip + 4 < end) && ip_in_bb (cfg, cfg->cbb, next_ip) &&
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Not yet converted but could/should be.

File: mono/mini/method-to-ir.c [hunk 6]
```diff
@@ -12587,8 +12621,8 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				/* 
 				 * Optimize the common case of ldftn+delegate creation
 				 */
-				if ((sp > stack_start) && (ip + 6 + 5 < end) && ip_in_bb (cfg, cfg->cbb, ip + 6) && (ip [6] == CEE_NEWOBJ)) {
-					MonoMethod *ctor_method = mini_get_method (cfg, method, read32 (ip + 7), NULL, generic_context);
+				if ((sp > stack_start) && (next_ip + 4 < end) && ip_in_bb (cfg, cfg->cbb, next_ip) && (next_ip [0] == CEE_NEWOBJ)) {
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Not yet converted but could/should be.

File: mono/mini/method-to-ir.c [hunk 7]
```diff
@@ -12656,8 +12691,9 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				/*
 				 * Optimize the common case of ldvirtftn+delegate creation
 				 */
-				if ((sp > stack_start) && (ip + 6 + 5 < end) && ip_in_bb (cfg, cfg->cbb, ip + 6) && (ip [6] == CEE_NEWOBJ) && (ip > header->code && ip [-1] == CEE_DUP)) {
-					MonoMethod *ctor_method = mini_get_method (cfg, method, read32 (ip + 7), NULL, generic_context);
+				if (previous_il_op == CEE_DUP && (sp > stack_start) && (next_ip + 4 < end) && ip_in_bb (cfg, cfg->cbb, next_ip) && (next_ip [0] == CEE_NEWOBJ)) {
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Not yet converted but could/should be.

File: mono/mini/method-to-ir.c [hunk 8]
```diff
@@ -68,6 +68,7 @@
 #include <mono/utils/mono-threads-coop.h>
 #include <mono/utils/mono-utils-debug.h>
 #include <mono/utils/mono-logger-internals.h>
+#include <mono/metadata/verify-internals.h>
```
Comment (kumpera) [inline mono/mini/method-to-ir.c]: I'm sorta in the fence on exposing the verifier to the JIT more than we need to.
Is this due to the header shenanigans you added there?

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: Yes probably every label can be a function. Want that now?
But I don't see the problem. goto and labels are a basic feature.

File: mono/mini/method-to-ir.c (hunk 8 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: It was to reuse the macros for token to index, token to  table -- token to fielddef.

File: mono/metadata/opcodes.h (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/opcodes.h]: Some combination of things.
Public header to use internal enum I think.

File: mono/mini/method-to-ir.c (hunk 8 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: It was just  for IS_FIELD_DEF I think.

**Comment by jaykrell** [comment]
Close and reopen to retry dotnet cla.

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: Actually the labels can probably be removed by first switching on operand type in the opcode table. Then the labels will all be adjacent.

**Comment by jaykrell** [comment]
Ok to merge once CI clears? I reset CI to get license to reattempt.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Pull request merged**